### ISSUE TITLE
Are there any ways to optimize inflow locations?

### ISSUE THREAD

**Issue by somisawa** [tum-pbs/PhiFlow#95]
In [this demo](https://tum-pbs.github.io/PhiFlow/Fluids_Tutorial.html), it is explained to obtain grad of grid values.

However, I'd like to optimize the center location of inflow like this:
```python
# informal code
inflow_location = tensor(...)
initial_inflow = CenteredGrid(Sphere(inflow_location, ...), ...)

inflow_location_grad = # I want this

inflow_location -= 1e-2 * inflow_location_grad
```

Is it possible?

**Comment by holl-** [comment]
Yes, you can optimize any input to a simulation. You just need to make it a parameter of your simulation / objective function.

```python
def simulate(smoke: CenteredGrid, velocity: StaggeredGrid, inflow: CenteredGrid):
  for _ in range(20):
    smoke = advect.mac_cormack(smoke, velocity, dt=1) + inflow
    buoyancy_force = smoke * (0, 0.5) @ velocity
    velocity = advect.semi_lagrangian(velocity, velocity, dt=1) + buoyancy_force
    velocity, _ = fluid.make_incompressible(velocity)
  loss = math.sum(field.l2_loss(diffuse.explicit(smoke - field.stop_gradient(smoke.inflow_loc[-1]), 1, 1, 10)))
  return loss, smoke, velocity

sim_grad = field.functional_gradient(simulate, wrt='inflow', get_output=False)

inflow_location_grad = sim_grad(...)
```

**Comment by holl-** [comment]
Right, the returned gradient will be for the inflow field. It's a scalar field so the gradient will also be a scalar field (CenteredGrid) of the same shape.
